src/common/imageio_libraw.c

```diff
@@ -0,0 +1,198 @@
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifdef HAVE_LIBRAW
+#include "common/darktable.h"
+#include "imageio.h"
+#include "imageio_gm.h"
+#include "develop/develop.h"
+#include "common/exif.h"
+#include "common/colorspaces.h"
+#include "control/conf.h"
+
+#include <memory.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include <strings.h>
+#include <assert.h>
+
+#include <libraw/libraw.h>
+
+
+/* LibRAW is expected to read only new canon CR3 files */
+
+static gboolean _supported_image(const gchar *filename)
+{
+  const char *extensions_whitelist[] = { "cr3", NULL };
+  gboolean supported = FALSE;
+  char *ext = g_strrstr(filename, ".");
+  if(!ext) return FALSE;
+  ext++;
+  for(const char **i = extensions_whitelist; *i != NULL; i++)
+    if(!g_ascii_strncasecmp(ext, *i, strlen(*i)))
+    {
+      supported = TRUE;
+      break;
+    }
+  return supported;
+}
+
+
+
+gboolean dt_libraw_lookup_makermodel(const char *maker, const char *model,
+                                     char *mk, int mk_len, char *md, int md_len,
+                                     char *al, int al_len)
+{
+  gboolean got_it_done = FALSE;
+
+  if(g_str_equal(maker, "Canon"))
+  {
+    g_strlcpy(mk, "Canon", mk_len);
+
+    if(g_str_equal(model, "Canon EOS RP"))
+    {
+      g_strlcpy(md, "EOS RP", md_len);
+      g_strlcpy(al, "EOS RP", al_len);
+    }
+    if(g_str_equal(model, "Canon EOS R"))
+    {
+      g_strlcpy(md, "EOS R", md_len);
+      g_strlcpy(al, "EOS R", al_len);
+    }
+    got_it_done = TRUE;
+  }
+
+  return got_it_done;
+}
+
+
+dt_imageio_retval_t dt_imageio_open_libraw(dt_image_t *img, const char *filename, dt_mipmap_buffer_t *mbuf)
+{
+  int err = DT_IMAGEIO_FILE_CORRUPTED;
+  int libraw_err = 0;
+  if(!_supported_image(filename)) return DT_IMAGEIO_FILE_CORRUPTED;
+  if(!img->exif_inited) (void)dt_exif_read(img, filename);
+
+  libraw_data_t *raw = libraw_init(0);
+  if(!raw) return DT_IMAGEIO_FILE_CORRUPTED;
+
+  libraw_err = libraw_open_file(raw, filename);
+  if(libraw_err != LIBRAW_SUCCESS) goto error;
+
+  libraw_err = libraw_unpack(raw);
+  if(libraw_err != LIBRAW_SUCCESS) goto error;
+
+  // Bad method to detect if camera is fully supported by libraw.
+  // But seems to be the best available. libraw crx decoder can actually
+  // decode the raw data, but internal metadata like wb_coeffs, crops etc.
+  // are not populated into libraw structure.
+  if(raw->color.cam_mul[0] == 0.0 || isnan(raw->color.cam_mul[0]))
+  {
+    libraw_close(raw);
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+
+  libraw_err = libraw_dcraw_process(raw);
```

Comment:
Why do you need to fully process (i.e. develop) the raw here, are the necessary raw metadata fields not populated after just unpack?

Reply:
Without libraw_dcraw_process() the image is black and iirc WB was not populated.

Reply:
Hm, that's kinda not ideal, because one end up doing tons of processing for nothing - imagine large batch imports!

Not the behaviour I remember when using rawpy (a LibRaw wrapper)... I'll test it again on some CR3s this time.

Reply:
Just tried on the Canon M50 sample, and this minimal example:

```
with rawpy.imread(path) as raw:
    imageio.imwrite('cr3_raw_image.tif', raw.raw_image)
    print(raw.camera_whitebalance)
    print(raw.daylight_whitebalance)
```

works with no processing (rawpy.imread just does unpack): raw data is not zero, and both as shot white balance and daylight white balance are not zero...

Does the behaviour change form model to model, or whether it's CRAW?

Reply:
Nope, I've commented the call out and any image in dt was just black afterwards. From my understanding unpack should be enough but the libraw API is somewhat special...